pkg/phlaredb/head.go

```diff
@@ -348,10 +349,29 @@ func (h *Head) Ingest(ctx context.Context, p *profilev1.Profile, id uuid.UUID, e
 
 	var profileIngested bool
 	for idxType := range samplesPerType {
+		samples := samplesPerType[idxType]
+		sort.Slice(samples, func(i, j int) bool {
+			return samples[i].StacktraceID > samples[j].StacktraceID
+		})
+		total := len(samples)
+		samples = slices.RemoveInPlace(samples, func(s *schemav1.Sample, i int) bool {
+			if s.Value == 0 {
+				return true
+			}
+			if i < len(p.Sample)-1 && s.StacktraceID == samples[i+1].StacktraceID {
+				samples[i+1].Value += s.Value
+				return true
+			}
+			return false
+		})
+		if total != len(samples) {
+			// copy samples if there are less than received to avoid retaining memory.
+			samples = copySlice(samples)
+		}
```

Comment:
We avoid copy if nothing is changed, but we make a copy if we have removed sample. This is because we keep those sample in memory until we flush and the less the better. Re-slicing keep reference to all samples even the one not accessible anymore.